README.md

```diff
@@ -1 +1,74 @@
-# depub
+# depub: minimise visibility
+
+## Overview
+
+When working on medium or large sized Rust code bases, it can be hard to know
+whether the visibility of functions, structs, and so on are still at the
+minimum required. For example, sometimes functions that once needed to be `pub`
+now only need to be `pub(crate)`, `pub(super)`, or simply private.
+
+`depub` minimises the visibility of such items in files passed to it, using a
+user-specified command (e.g. `cargo check`) as an oracle to tell if its
+reduction of an item's visibility is valid or not. In essence, `depub` does a
+string search for `pub`, replaces it with `pub crate` and sees if a test
+command still succeeds. If it does, it keeps that visibility, otherwise it
+replaces with the original and tries the next item. Note that `depub` is
+inherently destructive: it overwrites files as it operates, so do not run it on
+source code that you do not want altered!
+
+The list of visibilities that `depub` considers is, in order: `pub`,
+`pub(crate)`, `pub(super)`, and private (i.e. no `pub` keyword at all). `depub`
+searches for `pub`/`pub(crate)`/`pub(super)` instances, reduces their
+visibility by one level, and tries the oracle command. If it succeeds, it tries
+the next lower level until private visibility has been reached.
+
+Since reducing the visibility of one item can enable other items' visibility to
+be reduced, `depub` keeps running "rounds" until a fixed point has been
+reached. The maximum number of rounds is equal to the number of visible items
+in the code base, though in practise 2 or 3 rounds are likely to be all that is
+needed.
+
+
+## Usage
+
+`depub`'s usage is as follows:
+
+```
+depub -c <command> file_1 [... file_n]
+```
+
+where `<command>` is a string to be passed to `/bin/sh -c` for execution to
+determine whether the altered source code is still valid.
+
+To reduce the visibility of a normal Rust project, `cd` to your Rust code base
+and execute:
+
+```
+$ find . -name "*.rs" | \
```

Comment:
This is begging for a cargo plugin to hide the file search from users. Just as `cargo fmt` finds and formats all of the rust files in a project, perhaps`cargo depub` could find and depub all of the rust files in a project. Maybe later?

Reply:
Way too much complexity for now! Also, at least at the moment, it's sometimes useful to only pass a subset of files.